### PR TITLE
Lab 12

### PR DESCRIPTION
**Pull request by zilinbian56**
Standard deep learning and neural networks have given astounding performances on many tasks, but they commonly do not capture model uncertainty. In some cases of prediction, however, it is quite important to tell how confident we are about the model estimation, not just the the prediction result. To achieve this goal, we utilize Bayesian neural networks to get a predictive distribution of results instead of point estimation in this section. Specifically, we walk through the process of enabling dropout and getting the estimation of the predictive distribution. Then, we perform experiments using Bayesian neural networks with different non-linearities.
--Lab12 (23 Apr 2019) by Zilin Bian, Di Sha. Edited by Jong Yeob Kim

### CODE REVIEW

**Comment by Atcold** [comment]
Plz, edit the first message, similarly to how it's done in #1.
Also, make the summary more generic, explaining what has been covered in the lesson and why, rather than listing the chapters that this PR is introducing.

**Comment by zilinbian56** [comment]
> Plz, edit the first message, similarly to how it's done in #1.
> Also, make the summary more generic, explaining what has been covered in the lesson and why, rather than listing the chapters that this PR is introducing.

Hi there,
We've changed the summary and updated it in the first message, please have a look at it.

**Comment by Atcold** [comment]
If you check #1's first post, you'll see something like:

> -- Second part of Lecture 1 (28 Jan 2019) by Erica Dominic and Hongyu (Florence) Lu. Edited by Michael Gold

Add this missing line to your first post. Thanks.

File: labs/12/theory.tex [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+\begin{equation}
+    Var(y^*) \approx \tau^{-1} I_D + \frac{1}{T}\sum^{T}_{t=1} \hat{y_t}^*(x^*)^T \hat{y_t}^* (x^*) - \mathbb{E} (y^*)^T \mathbb{E} (y^*)
+\end{equation}
+\\
+Eq. 17.1 was introduced as model averaging and it was explained that scaling the weights at test time without dropout gives a reasonable approximation to this equation. Eq. 17.2 is simply the sample variance of $T$ forward passes through the network plus the inverse model precision. Note that the vectors above are row vectors and that the products are outer-products.
+\\
+Implementation in python code to get the predictive mean and uncertainty is easy:
+\begin{figure}[H]
+    \centering
+    \includegraphics[width=\textwidth]{labs/12/images/python code of getting mean and var.PNG}
```
Comment (Atcold) [inline labs/12/theory.tex:28]: Replace this image with the code itself. Thanks.
See other PRs for how this has been done, with the minted package.

File: labs/12/theory.tex [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    \caption{Python code to obtain predictive mean and uncertainty from dropout}
+    \label{fig:visual_caption}
```
Comment (Atcold) [inline labs/12/theory.tex:30]: `visual_caption` is already defined here https://github.com/Atcold/SP19-DL-collaborative-notes/blob/master/labs/07/theory.tex#L14

Comment (zilinbian56) [reply labs/12/theory.tex:30]: Please check PR Lab12 #53

File: labs/12/theory.tex (hunk 1 above)
Comment (zilinbian56) [reply labs/12/theory.tex:28]: Please check PR Lab12 #53

**Pull request merged**